gateway/src/test/java/io/zeebe/gateway/broker/BrokerClientBackPressureTest.java

```diff
@@ -0,0 +1,142 @@
+import io.zeebe.gateway.impl.configuration.GatewayCfg;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceCreationRecord;
+import io.zeebe.protocol.record.ValueType;
+import io.zeebe.protocol.record.intent.JobBatchIntent;
+import io.zeebe.protocol.record.intent.WorkflowInstanceCreationIntent;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import io.zeebe.test.broker.protocol.brokerapi.ExecuteCommandResponseBuilder;
+import io.zeebe.test.broker.protocol.brokerapi.StubBrokerRule;
+import io.zeebe.test.util.AutoCloseableRule;
+import io.zeebe.transport.impl.util.SocketUtil;
+import io.zeebe.util.sched.clock.ControlledActorClock;
+import java.util.concurrent.ExecutionException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TestName;
+
+public class BrokerClientBackPressureTest {
+
+  private static final int CLIENT_MAX_REQUESTS = 128;
+  @Rule public StubBrokerRule broker = new StubBrokerRule();
+  @Rule public AutoCloseableRule closeables = new AutoCloseableRule();
+  @Rule public ExpectedException exception = ExpectedException.none();
+  @Rule public TestName testContext = new TestName();
+  private BrokerClient client;
+  private ControlledActorClock clock;
+
+  @Before
+  public void setUp() {
+    final GatewayCfg configuration = new GatewayCfg();
+    configuration
+        .getCluster()
+        .setHost("0.0.0.0")
+        .setPort(SocketUtil.getNextAddress().port())
+        .setContactPoint(broker.getSocketAddress().toString())
+        .setRequestTimeout("3s");
+    configuration
+        .getBackpressure()
+        .setEnabled(true)
+        .getAimdCfg()
+        .setInitialLimit(1)
+        .setMinLimit(1)
+        .setMaxLimit(1);
+    clock = new ControlledActorClock();
+
+    final AtomixCluster atomixCluster = mock(AtomixCluster.class);
+    final ClusterMembershipService memberShipService = mock(ClusterMembershipService.class);
+    when(atomixCluster.getMembershipService()).thenReturn(memberShipService);
+
+    client = new BrokerClientImpl(configuration, atomixCluster, clock);
+
+    ((BrokerClientImpl) client).getTransport().registerEndpoint(0, broker.getSocketAddress());
+
+    final BrokerClusterStateImpl topology = new BrokerClusterStateImpl();
+    topology.addPartitionIfAbsent(START_PARTITION_ID);
+    topology.setPartitionLeader(START_PARTITION_ID, 0);
+
+    ((BrokerTopologyManagerImpl) client.getTopologyManager()).setTopology(topology);
+  }
+
+  @After
+  public void tearDown() {
+    client.close();
+  }
+
+  @Test
+  public void shouldReturnErrorOnBackpressure() {
+    // given
+    final BrokerCreateWorkflowInstanceRequest request = new BrokerCreateWorkflowInstanceRequest();
+    request.setPartitionId(1);
+    client.sendRequest(request);
+
+    // then
+    exception.expect(ExecutionException.class);
```

Comment:
Can we use `assertThatThrownBy` (or whatever the name is) instead of the expected exception?

Reply:
Also aren't we returning a better exception than `ExecutionException`? In the cause I guess it should be `ResourceExhausedException`? I'd prefer matching on the exception type than the message, if possible.

Reply:
Matching with `ResourceExhaustedException` did not work. I didn't check at what point it is converted to ExecutionException.